### PR TITLE
Add Buffered BinaryReader to speed up Binlog reading

### PR DESCRIPTION
**Pull request by yuehuang010**
# Improve Binlog Reading by using a Custom Binary Reader 
- Added Buffered-Binary-Reader that combines MemoryStream and BinaryReader. 
- This allows `Read7BitEncodedInt()` to inline `Stream.ReadByte()`, saving 1-5 virtual function calls per int reads.
- Added BulkRead7BitEncodedInt() to further optimize multiple read operations to further save on memory allocations.
- These changes resulted in 2-4X speed up when tested reading a very large binlog.
- Added a IBinaryReader interface to make switching back to BCL BinaryReader easy.
# Improve Binlog Writer's Write7IntEncodedInt().
- Write to a buffer before writing to the stream.  Reduce from 1 - 5 virtual function calls to 1.
- This change reduces the total time in the function, but since it is not on a critical path, its benefit is negligible.

### CODE REVIEW

File: src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        int IBinaryReader.BytesCountAllowedToReadRemaining => _readStream.BytesCountAllowedToReadRemaining;
+
+        int? IBinaryReader.BytesCountAllowedToRead { set => _readStream.BytesCountAllowedToRead = value; }
+
+        long IBinaryReader.Position => _readStream.Position;
+
+        public const int MaxBulkRead7BitLength = 10;
+        private int[] resultInt = new int[MaxBulkRead7BitLength];
+
+        int[] IBinaryReader.BulkRead7BitEncodedInt(int numIntegers)
```
Comment (JanKrivanek) [inline src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs:44]: We should probably check if the arg doesn't go over the `MaxBulkRead7BitLength` (and then either throw or just read at max `MaxBulkRead7BitLength`)

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+            int readChunk = 0;
+
+            if (stringLength == 0)
+            {
+                return string.Empty;
+            }
+
+            if (stringLength < 0)
+            {
+                throw new Exception();
```
Comment (JanKrivanek) [inline src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:115]: let's use specific type and message (couple other cases in this file as well)

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+        private StringBuilder? cachedBuilder;
+
+        // Reusable char[] for ReadString().
+        private char[]? charBuffer;
+
+        /// <summary>
+        /// Reads a string with a prefixed of the length.
+        /// </summary>
+        /// <returns>A string.</returns>
+        public string ReadString()
```
Comment (JanKrivanek) [inline src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:102]: Where is this implementation comming from? Is it adapted from the BCL BinaryReader? We should have that info in comments. I already saw couple cases where MSBuild code had some custom implementation copied from BCL - while the BCL diverged a bit over time and nobody knew the original motivation or sometime even the source

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+            string result = cachedBuilder.ToString();
+            cachedBuilder.Clear();
+            return result;
+        }
+
+        /// <summary>
+        /// Reads an 8-byte signed integer.
+        /// </summary>
+        /// <returns></returns>
+        public long ReadInt64()
```
Comment (JanKrivanek) [inline src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:168]: Same here - where is it comming from? Is Big/Little Endian handled accordingly?

**Review by JanKrivanek — COMMENTED** [review]
Couple comments.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs (hunk 4 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:166]: Added note that it is from BCL.  I will need to check with the Endian, I believe the test will catch the issue on MacOS.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs (hunk 3 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:104]: ReadString() is effectively re-written when compared to BCL.  Should review in full.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs (hunk 2 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs]: Switched to `throw new FormatException();`

File: src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs (hunk 1 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs:44]: Check added.

**Review by JanKrivanek — COMMENTED** [review]
Added coupe comments.

Let us know once ready for full review (Perf comaprison would be nice then)

File: src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs (hunk 1 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs:44]: Oh, I added to the buffered reader but missed this one.

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 5]
```diff
@@ -1033,15 +1034,34 @@ private void WriteProperties(IEnumerable properties)
...
+            foreach(int num in ints)
+            {
+                uint v = (uint)num;   // support negative numbers
+                while (v >= 0x80)
+                {
+                    buildEventContextBuffer[index++] = (byte)(v | 0x80);
+                    v >>= 7;
+                }
+
+                buildEventContextBuffer[index++] = (byte)v;
```
[the remark this answers is not in the document]
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:1062]: I modify this function to write to a byte array before writing to the stream.  This reduces 7 stream writes to 1 write.

This using 7bit encoding while the calls in BinaryWriterExtensions is using 4bytes Int32.  These two function are not compatible with each other.  You might have more history as to why this is the case.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+        /// <returns>An array of Integers with the results.</returns>
+        /// <remarks>This will reuse the same array for results to avoid extra allocations.</remarks>
+        public int[] BulkRead7BitEncodedInt(int numIntegers)
+        {
+            if (numIntegers > MaxBulkRead7BitLength)
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+
+            FillBuffer(5 * numIntegers, throwOnEOF: false);
```
[the remark this answers is not in the document]
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:311]: Add comment about `5`.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+        /// <summary>
+        /// If <see cref="BytesCountAllowedToRead"/> is set, then this is the number of bytes remaining to read.  Otherwise, 0.
+        /// </summary>
+        public int BytesCountAllowedToReadRemaining => maxAllowedPosition == long.MaxValue ? 0 : (int)(maxAllowedPosition - baseStreamPosition);
+
+        /// <summary>
+        /// Reads a 32-bit signed integer.
+        /// </summary>
+        /// <returns>Return a integer.</returns>
+        /// <remarks>Logic copied from BCL BinaryReader.</remarks>
```
[the remark this answers is not in the document]
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs]: Added.

**Comment by yuehuang010** [comment]
> Perf comparison would be nice then

I ported these changes to a private [BinLogViewer](https://github.com/yuehuang010/MSBuildStructuredLog/tree/dev/yuehuang/main_bufferedreader_port) to open a 500mb binlog.

Before:
Total opening time: 1:11.670, **Loading: 51.482 s**, Analyzing: 8.743 s, Indexing: 7.944 s, Allocated: 120,133,521,336 bytes, Nodes: 143,137,731, Strings: 3,903,327

After:
Total opening time: 52.185 s, **Loading: 31.396 s**, Analyzing: 9.339 s, Indexing: 7.796 s, Allocated: 63,378,355,056 bytes, Nodes: 143,137,731, Strings: 3,903,327

Note: This removes the bottleneck and exposes other places for further optimize.  The gains are much greater in unit testing than End to End.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs (hunk 3 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:107]: I modified to look closer to the version in BCL.  It is very similar, I moved the one time branch outside of the loop to simplified the loop.

File: src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs (hunk 4 above)
Comment (yuehuang010) [reply src/Build/Logging/BinaryLogger/BufferedBinaryReader.cs:169]: Looks like endian doesn't matter as it is all little endian regardless of MacOS.  The binary writer should need be mindful to convert to little endian.

**Pull request closed without merging**